packages/babel-types/scripts/utils/stringifyValidator.js

```diff
@@ -1,40 +1,55 @@
+// @ts-check
+/**
+ * Stringify a validator to its corresponding TypeScript type.
+ * @param {import("../../src/definitions/utils.ts").Validator | undefined} validator
+ * @param {string} nodePrefix
+ * @returns {string}
+ */
 export default function stringifyValidator(validator, nodePrefix) {
   if (validator === undefined) {
     return "any";
   }
 
-  if (validator.each) {
+  if ("each" in validator) {
     return `Array<${stringifyValidator(validator.each, nodePrefix)}>`;
   }
 
-  if (validator.chainOf) {
-    const ret = stringifyValidator(validator.chainOf[1], nodePrefix);
-    return Array.isArray(ret) && ret.length === 1 && ret[0] === "any"
-      ? stringifyValidator(validator.chainOf[0], nodePrefix)
-      : ret;
+  if ("chainOf" in validator) {
```

Comment:
Previously the `chainOf` depends on the special return value `["any"]` for general validator `ValidatorImpl`, here we always return `any` for such validators and then extract the first meaningful type when iterating from the end to start.